### ISSUE TITLE
Can `MambaForCausalLM` be used directly for training instead of `AutoModelForCausalLM`?

### ISSUE THREAD

**Issue by LumenScopeAI** [state-spaces/mamba#294]
Hello,

I'm currently working with the `transformers` library to train a model on causal language modeling tasks using the `MambaForCausalLM` class. However, I've noticed that the typical approach to training in the library uses `AutoModelForCausalLM` to load the model for training, and I'm wondering if it's possible and recommended to use `MambaForCausalLM` directly for training instead.

Here is the code snippet I'm referring to:

```python
from datasets import load_dataset
from trl import SFTTrainer
from peft import LoraConfig
from transformers import AutoTokenizer, AutoModelForCausalLM, TrainingArguments
tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba-370m-hf")
model = AutoModelForCausalLM.from_pretrained("state-spaces/mamba-370m-hf")
dataset = load_dataset("Abirate/english_quotes", split="train")
training_args = TrainingArguments(
    output_dir="./results",
    num_train_epochs=3,
    per_device_train_batch_size=4,
    logging_dir='./logs',
    logging_steps=10,
    learning_rate=2e-3
)
lora_config =  LoraConfig(
        r=8,
        target_modules=["x_proj", "embeddings", "in_proj", "out_proj"],
        task_type="CAUSAL_LM",
        bias="none"
)
trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    args=training_args,
    peft_config=lora_config,
    train_dataset=dataset,
    dataset_text_field="quote",
)
trainer.train()

```

In inference, I successfully use `MambaForCausalLM` as follows:

```python
from transformers import MambaConfig, MambaForCausalLM, AutoTokenizer
import torch

tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba-370m-hf")
model = MambaForCausalLM.from_pretrained("state-spaces/mamba-370m-hf")
input_ids = tokenizer("Hey how are you doing?", return_tensors="pt")["input_ids"]

out = model.generate(input_ids, max_new_tokens=10)
print(tokenizer.batch_decode(out))

```

Could you clarify if using `MambaForCausalLM` for training is supported and if there are any additional configurations required for this?

Thank you for your assistance.

**Comment by tridao** [comment]
Sorry we're not familiar with how huggingface's `transformers` implement Mamba.

**Comment by vasqu** [comment]
They lead to the same class being used, `AutoModelForCausalLM` is basically just a wrapper that infers your specific model, e.g. mamba, for the case, e.g. causal lm, (by using the config of a model iirc, see [mamba-130m config](https://huggingface.co/state-spaces/mamba-130m-hf/blob/main/config.json)).

Minimal snippet to show it leads to the same:
```
from transformers import AutoModelForCausalLM, MambaForCausalLM

mamba_model_auto = AutoModelForCausalLM.from_pretrained('state-spaces/mamba-130m-hf')
mamba_model_explicit = MambaForCausalLM.from_pretrained('state-spaces/mamba-130m-hf')

print(f'Auto type - {type(mamba_model_auto)} | Explicit type - {type(mamba_model_explicit)}')
```
This should print something like `Auto type - <class 'transformers.models.mamba.modeling_mamba.MambaForCausalLM'> | Explicit type - <class 'transformers.models.mamba.modeling_mamba.MambaForCausalLM'>`.

tl;dr: it doesn't matter, you can use both in any case
